WIP Shopping List Card

iantrich opened the pull request:
There are two issues I'm still trying to work through:

1. I'm using a mdi icon for the clear-all-notifications. I could not get a hass version to work.
2. I've been unsuccessful in trying to use `localize` for the 'Add Item' text. There is already translations for this text, but perhaps I was not properly calling `localize`

iantrich committed:
WIP Shopping List Card

balloob commented:
When adding a new hass icon, restart `script/develop`

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,217 @@
+  }
+
+  public getCardSize() {
+    return 1;
```
Doesn't seem right?

zsarnett replied:
Needs to be calculated on load but what if it grows. then on reload the positioning would change right?

balloob replied:
Yeah, we call it only once. We should set it to 3 as that seems like a good average size?

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,217 @@
+          ></paper-input>
+          <paper-icon-button
+            slot="item-icon"
+            icon="hass:plus"
```
`.icon`

iantrich replied:
doesn't work

balloob commented:
Please include a screenshot and a demo for the gallery

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,217 @@
+          ></paper-icon-button>
+        </div>
+
+      ${Object.values(this.items).map(item =>
```
items is an array. Don't need object values?

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,217 @@
+        <paper-icon-item>
+          <paper-checkbox
+            slot="item-icon"
+            ?checked=${item.complete}
```
We should use `.checked`

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,217 @@
+          <paper-item-body>
+            <paper-input
+              id='editBox'
+              no-label-float
```
`.noLabelFloat`

iantrich replied:
doesn't work

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,217 @@
+            <paper-input
+              id='editBox'
+              no-label-float
+              value='${item.name}'
```
`.value`

iantrich replied:
this one does work 🤷‍♂️

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,217 @@
+
+  private connectedCallback() {
+    super.connectedCallback();
+    this._fetchData = this._fetchData.bind(this);
```
Why?

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,217 @@
+  }
+
+  private _fetchData() {
+    this.hass.callApi("GET", "shopping_list").then(
```
We can't just start copy/pasting all these methods from file to file… that's unmaintainable.

Create a new folder `src/data`. In that folder we're going to create a new folder for each component. So create `src/data/shopping_list/data.js` and add 
```js
export fetchItems = (hass) => hass.callApi("GET", "shopping_list");
```

Do not include the `reverse` call in `data.js`, that's an implementation detail of this card.

Do this for all API calls.

iantrich replied:
does it make sense to create a folder for each component to house just a single file in it?

why not just `src/data/shopping_list.js` or `src/data/components/shopping_list.js`?

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,217 @@
+  private _itemCompleteTapped(ev, id) {
+    ev.stopPropagation();
+    this.hass
+      .callApi("POST", "shopping_list/item/" + id, {
```
Extract too. `export const completeItem(hass, itemId) => hass.callApi(…)`

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,217 @@
+  }
+
+  private _addItem(ev) {
+    if (!this.shadowRoot.querySelector('#addBox').value) {
```
When can this happen?

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,217 @@
+      return;
+    }
+
+    this.hass.callApi("POST", "shopping_list/item", {
```
Extract

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,217 @@
+  }
+
+  private _saveEdit(ev, id) {
+    this.hass.callApi("POST", "shopping_list/item/" + id, {
```
extract

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,217 @@
+  }
+
+  private _clearCompleted() {
+    this.hass.callApi("POST", "shopping_list/clear_completed");
```
extract

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,217 @@
+        <div class="addRow">
+          <paper-input
+            id='addBox'
+            placeholder='Add Item'
```
`.placeholder`

When you're rendering a web component, prefer properties over attributes.

balloob replied:
Except for when you need it for aria or CSS, things that require attributes

iantrich replied:
Out of curiosity do you have any references on why we should prefer properties to attributes? Would like to read more on any advantages or the reasons behind it. Thanks.

iantrich replied:
`.placeholder` doesn't work either

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,217 @@
+      return html``;
+    }
+
+    // TODO Localize 'Add Item' text
```
Use `this.localize`

iantrich replied:
`placeholder='${this.localize('ui.panel.shopping-list.add_item')}'` does not return a value. No errors in console though

frenck commented:
Matching documentation PR is missing for this PR. Please be sure to update the documentation and create a PR on our docs repository as well.

iantrich commented:
> When adding a new hass icon, restart `script/develop`

@balloob `gen-icons.js` only processes `.js` files, not `.ts` files. Should I put that in a different PR to fix?

iantrich committed:
Extract API calls out for shopping-list

iantrich committed:
Address review comments

connectedCallback is not working
`notification-clear-all` icon is not working until `gen-icons.js` is updated to include `.ts` extensions
`hass.connection` is throwing TSLint errors as `type.ts` has is defined as `hass.conn`

iantrich commented:
@balloob `types.ts` has the `conn` instead of `connection` for type `HomeAssistant` which was giving me some issues as well. I'm going commit what I have for now and submit a PR for `gen-icons.js` and `types.ts`

If someone could also please look at my updated `connectedCallback` function. It doesn't work and it's late so won't be getting to it right now, so if anyone has a moment to see how I messed up, I would appreciate. The previous commit was working, but TS linting did not like the use of unsub, and I really just guessed.

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,212 @@
+
+    super.connectedCallback();
+    this.hass.connection
+      .subscribeEvents(this._fetchData, "shopping_list_updated")
```
subscribeEvents resolves to an unsub function. The card should unsub when it is disconnected

iantrich replied:
I'm still lost as how this should work

iantrich replied:
Here's the code from `ha-panel-shopping-list.js`
```
  connectedCallback() {
    super.connectedCallback();
    this._fetchData = this._fetchData.bind(this);

    this.hass.connection
      .subscribeEvents(this._fetchData, "shopping_list_updated")
      .then(
        function(unsub) {
          this._unsubEvents = unsub;
        }.bind(this)
      );
    this._fetchData();
  }

  disconnectedCallback() {
    super.disconnectedCallback();
    if (this._unsubEvents) this._unsubEvents();
  }
```

I don't understand the `this._unsubEvents` portion

balloob replied:
So when we subscribe to events, we create a subscription for the current websocket connection. The backend will keep sending the events over the websocket connection as long as the subscription is around. So when we no longer need this component and remove it from the DOM (and we're notified by `disconnectedCallback` being called), we should remove that subscription, or else the backend will keep sending updates. Now imagine yoy refresh your lovelace config 10 times, we now have 10 subscriptions…

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,212 @@
+          <paper-icon-button
+            slot="item-icon"
+            icon="hass:plus"
+            @click='${(e) => this._addItem(e)}'
```
you can just do `this._addItem`. Lit Element will make sure the call context is set correctly automatically

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,212 @@
+          <paper-icon-button
+            slot="item-icon"
+            icon="hass:notification-clear-all"
+            @click='${() => this._clearItems()}'
```
same.

iantrich replied:
Changing `@click='${() => this._clearItems()}'` to `@click='${this._clearItems()}'` resulted in `this._clearItems()` being called on initialization.

balloob replied:
yeah, duh, because you are calling it ;-) 

Try

```html
<paper-icon-button
  @click='${this._clearItems}'
></paper-icon-button>
```

iantrich replied:
That's slick!

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,212 @@
+          <paper-checkbox
+            slot="item-icon"
+            ?checked=${item.complete}
+            @click='${(e) => this._itemCompleteTapped(e, item.id)}'
```
assign the item ID as a property on `<paper-checkbox>`, that way you can read it in your event listener and we don't have to create a one off function. [Example](https://github.com/home-assistant/home-assistant-polymer/blob/master/src/panels/lovelace/cards/hui-glance-card.ts#L149)

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,212 @@
+          ></paper-checkbox>
+          <paper-item-body>
+            <paper-input
+              id='editBox'
```
not used.

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,212 @@
+    `;
+  }
+
+  private _fetchData() {
```
You can use an async function instead of promises.

```js
private async _fetchData() {
  if (this.hass) {
    this._assignItems(await fetchItems(this.hass));
  }
}
```

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,212 @@
+  }
+
+  private _assignItems(items: {}) {
+    this.items = items as ListItem[];
```
Instead of typecasting here, you should set the type signature for the parameter : `_assignItems(items: ListItem[])`

balloob commented on src/data/shopping_list.ts:
```diff
@@ -0,0 +1,30 @@
+import { HomeAssistant } from "../types.js";
+
+export const fetchItems = (hass: HomeAssistant) =>
```
Make sure to add return items to all functions in this file (and define the types in this file). 

```ts
export const fetchItems = (hass: HomeAssistant): ListItem[] =>
```

balloob replied:
Make sure that all the types are also exported: `export interface ListItem {`

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,212 @@
+  }
+
+  public getCardSize() {
+    return (this.config ? (this.config.title ? 1 : 0) : 0) + (this.items ? this.items.length : 0);
```
let's default to size of 3 items if no items available.

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,212 @@
+
+  static get properties() {
+    return {
+      hass: {},
```
We should only list things in properties if we want to re-render based on data that is contained in it. This card doesn't need to have `hass` in here.

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,212 @@
+    return {
+      hass: {},
+      config: {},
+      items: [],
```
items doesn't need to be in here, nor is `[]` a valid options object for lit.

balloob replied:
options are generally only used when you need to serialize/deserialize things from attributes.

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,212 @@
+          ></paper-icon-button>
+        </div>
+
+      ${this.items!.map((item) => html`
```
Since each item has an ID and it can happen that rows show up in between, we should use the repeat directive.

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,212 @@
+  }
+
+  private _itemCompleteTapped(ev, id) {
+    ev.stopPropagation();
```
why is this needed?

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,212 @@
+
+  private _itemCompleteTapped(ev, id) {
+    ev.stopPropagation();
+    if (this.hass) {
```
This check is not needed, we know `hass` is available because we wouldn't have been able to fetch items to attach this event handler to begin with. So to tell TS you know better, pass `this.hass!` to `completeItem`

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,212 @@
+  }
+
+  private _addItem(ev) {
+    if (this.shadowRoot && this.shadowRoot.querySelector("#addBox")) {
```
We will always have a shadowRoot because `addItem` is triggered by something in the shadowRoot.

iantrich commented:
What are the steps to bring this branch inline with newer commits to master?
```
git checkout shopping-list-card
git fetch upstream
git rebase upstream/master
```

balloob commented:
yes, and then `git push -f`

iantrich committed:
Address most review comments

Still unclear how to use the `connectedCallback` and `localize` function is still not working

iantrich commented:
Still unclear how to use the `connectedCallback` and `localize` function is still not working

iantrich commented:
> Matching documentation PR is missing for this PR. Please be sure to update the documentation and create a PR on our docs repository as well.

@frenck thanks! I'll be sure to add some before this moves to being merged.

iantrich commented:
@cgarwood Do you have a label for something like `documentation-needed` so we know that this shouldn't progress until I finish creating proper documentation for the new card?

iantrich committed:
Replaced map with repeat and let LitElement define event context

Tried to get connectedCallback working but get the error `Uncaught (in promise) TypeError: Cannot set property '_unsubEvents' of undefined
    at hui-shopping-list-card.ts:66`

iantrich commented:
Still having issues with `connectedCallback` and `this.localize`.
`connectedCallback` is giving the error `Uncaught (in promise) TypeError: Cannot set property '_unsubEvents' of undefined at hui-shopping-list-card.ts:66`

`this.localize` is not giving any errors but is not returning the expected string. Do I have the right format for the passed string? It's is the same as how the shopping list panel was calling.

iantrich commented:
Reasoning for change to line 65: https://stackoverflow.com/questions/41944650/this-implicitly-has-type-any-because-it-does-not-have-a-type-annotation

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,229 @@
+      super.connectedCallback();
+      this.hass.connection
+        .subscribeEvents(this._fetchData, "shopping_list_updated")
+        .then(function(this: HuiShoppingListCard, unsub: () => Promise<void>) {
```
the reason `this` is undefined is because you're using `function` instead of a lambda (`=>` syntax). You should always prefer the arrow syntax, it has less overhead and it maintains the `this` of the context in which it is defined.

balloob replied:
are you sure? Can you make it `() => this._fetchData()` instead of passing in `this._fetchData` ?

iantrich replied:
Thanks! I don't think I would have figured that one out. works now!

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,229 @@
+        <div class="addRow">
+          <paper-input
+            id='addBox'
+            placeholder='${this.localize("ui.panel.shopping-list.add_item")}'
```
The reason this doesn't work, is because the shopping list panel resources have not been loaded. It's all splitted up to minimize the amount of data our user has to load on startup. You should define new strings just for lovelace.

Here shopping list is defined: https://github.com/home-assistant/home-assistant-polymer/blob/master/src/translations/en.json#L835

Just add a new key as a sibling for lovelace. Under that make a new one card and then card type and then "add_item". Just fill in English, it will be automatically uploaded to the translation platform and our translator community will do the rest.

iantrich committed:
Addressed review comments

balloob commented:
Opened a PR to allow mocking Rest API calls for your gallery demo. Added a shopping list example 👍 
 https://github.com/home-assistant/home-assistant-polymer/pull/1859

iantrich commented:
@balloob my `connectedCallback` implementation is still not working :(

iantrich committed:
Fix to event listener

iantrich commented:
> Opened a PR to allow mocking Rest API calls for your gallery demo. Added a shopping list example 👍
> #1859

@balloob thanks! I'll have to study this demo area and figure out how I'm going to pull of this demo as the demo should be reflective of the actual actions...i.e. adding/removing to the list

balloob commented:
Nah, those actions can be ignored in the demo, just render a card is a good start.

iantrich committed:
Added demo

Gallery was giving me issues, so I have not fully tested it yet and wanted to move on to some other tasks

balloob commented:
In the demo, the items don't load until I try to add something.

I think for UX, we can do better. We should explore some other material based designs and see what they do.

iantrich commented:
@balloob 
Looks to be an issue with the connectedCallback. Do I need to mock that out perhaps?

Agreed on the UX. It is pretty much a replica of the panel. Could also transition some pieces to mwc. They have buttons and checkboxes finalized.

iantrich committed:
Added delete icon

Dependent on change to `shopping-list` component change: https://github.com/home-assistant/home-assistant/pull/17888

balloob commented:
![image](https://user-images.githubusercontent.com/1444314/47616330-abbf9880-dabb-11e8-8267-5297a8784f3d.png)

(in screenshot, Harissa row is hovered by mouse)

iantrich commented:
I don’t like the “add item” being at the bottom

iantrich committed:
Cosmetic changes (missing functionality)

iantrich commented:
![image](https://user-images.githubusercontent.com/1287159/47618974-1fa37600-daa7-11e8-9d94-b9e6fa01f5c6.png)

Here's what I have so far

- [ ] Need backend call to update order
- [ ] Need to figure out drag/drop for reordering
- [ ] Need to figure out how to show move/delete on hover only

zsarnett commented:
Showing on hover is just making it hidden/display none at first and then on hover show.
```
target:hover {
//visible
}
```

zsarnett commented:
I dont think the drag and drop reorder needs to be in the MVP

iantrich committed:
Attempt at getting :hover working

iantrich commented:
> I dont think the drag and drop reorder needs to be in the MVP

I've pushed up what I'm trying to do but is not working as I expect it to

iantrich committed:
Remove attempt to override hover

zsarnett commented:
Just add the icons to the Row and hide them unless hovered
```
editrow:hover > icondelete {
   visibility: visible;
}

iantrich commented:
@zsarnett that worked! Thanks!

iantrich committed:
Show edit options only on hover

iantrich committed:
FAB button WIP

. Made title required as the list without it and a FAB looked pretty rough, but not 100% on that
* Will need to show an input-text when the plus is clicked to enter your item. (This design does make it a little slow to enter multiple items to your list)

iantrich commented:
FAB button partially implemented. As I noted in my commit, I'm not 100% sure about it. Looks nice, but takes away some ease of use for users wanting to add lots of item to a list in a row. Perhaps a way around this is after you hit the '+', the dialog comes up and has an "add" and a "add more" where "add more" doesn't close the dialog (same for hitting enter). And then if you hit enter a 2nd time without typing anything, the dialog will close...? Who's our UX expert? LOL

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,279 @@
+import "@polymer/paper-input/paper-input.js";
+import "@polymer/paper-item/paper-icon-item.js";
+import "@polymer/paper-item/paper-item-body.js";
+import "@polymer/paper-fab/paper-fab.js";
```
absolutely no fab. It draws all the attention on the screen. And material design guidelines are saying not to use it in cards.

zsarnett replied:
Revert, Revert

iantrich replied:
it didn't feel great while adding it :)

balloob replied:
import is still there.

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,279 @@
+        this._items!,
+        (item) => html`
+        <div class="editRow">
+          <paper-icon-button
```
The first version of this card will need to be simple, without requiring any backend changes or anything. Just as basic as it gets. If people want to clear completed, let them go to the shopping list panel.

balloob replied:
So let's drop drag, let's drop delete. Just show the existing items and allow checking them off. Maybe add a button at bottom of the card that is just `<a href='/shopping-list'><paper-button>Go to shopping list</paper-button></a>`

iantrich replied:
* I have all the functionality for adding/clearing completed already set.
* I'll remove the delete and drag until the back end can be updated

iantrich committed:
Revert FAB changes

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,279 @@
+          margin-left: 7px;
+        }
+        .fabDiv:hover > .moreOptions {
+          visibility: visible;
```
you almost never want to use `visibility` because the elements will keep occupying the space, they are just not visible. You can see this with the delete buttons on your card:

![image](https://user-images.githubusercontent.com/1444314/47620782-a29bef00-daee-11e8-8c77-02398ba203c1.png)

`display: none` is the correct CSS modifier.

iantrich committed:
MVP

balloob commented on src/data/shopping_list.ts:
```diff
@@ -0,0 +1,34 @@
+  hass: HomeAssistant,
+  itemId: number,
+  name: string
+): Promise<{}> =>
```
`Promise<void>`, as you don't expect any return value.

balloob replied:
actually, you get the shopping list item back. Did you look at the actual return values on the backend?

balloob commented on src/data/shopping_list.ts:
```diff
@@ -0,0 +1,34 @@
+  hass: HomeAssistant,
+  itemId: number,
+  complete: boolean
+): Promise<{}> =>
```
same

balloob commented on src/data/shopping_list.ts:
```diff
@@ -0,0 +1,34 @@
+    complete,
+  });
+
+export const addItem = (hass: HomeAssistant, name: string): Promise<{}> =>
```
idem

balloob replied:
Actually, this returns the JSON of the newly created item.

balloob commented on src/data/shopping_list.ts:
```diff
@@ -0,0 +1,34 @@
+    name,
+  });
+
+export const clearItems = (hass: HomeAssistant): Promise<{}> =>
```
yep

balloob commented on src/data/shopping_list.ts:
```diff
@@ -0,0 +1,34 @@
+export const clearItems = (hass: HomeAssistant): Promise<{}> =>
+  hass.callApi("POST", "shopping_list/clear_completed");
+
+export const deleteItem = (hass: HomeAssistant, itemId: number): Promise<{}> =>
```
oh oh

balloob commented on src/data/shopping_list.ts:
```diff
@@ -0,0 +1,34 @@
+import { HomeAssistant } from "../types.js";
+import { ShoppingListItem } from "../panels/lovelace/cards/hui-shopping-list-card.js";
```
…

The fact that a generic data folder needs to import something specific from a Lovelace card should have been a pretty red flag. 

The types for shopping list should be in this file.

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,236 @@
+          ></paper-icon-button>
+        </div>
+
+      ${repeat(
```
You should add a keyFn as second arg: `item => item.id` [docs](https://polymer.github.io/lit-html/guide/writing-templates#repeatitems-keyfn-template)

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,236 @@
+              no-label-float
+              .value='${item.name}'
+              .itemId=${item.id}
+              @change=${this._saveEdit}
```
This will fire A LOT of events. Every key press will fire a change event, that's then being sent to the server, changing an item generates an event, event is acted upon by this card by fetching items again. Then the user presses another key…

iantrich replied:
It only fires when we lose focus

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,236 @@
+        }
+        paper-input {
+          --paper-input-container-underline: {
+            display: none;
```
Why? It's not clear that I can click on it and start typing.

balloob replied:
It looks more like it's broken and something didn't render

![image](https://user-images.githubusercontent.com/1444314/47621492-16db9000-daf9-11e8-8ffc-c97bbffdcfe9.png)

iantrich replied:
There should be a placeholder there and there is on mine. I've been having strange issues with the paper-input in general for the addBox and not sure why.

iantrich replied:
Something to do with my use of localize?

balloob replied:
Ah, that should resolve itself when we merge then.

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,236 @@
+      this._items.reverse();
+    }
+
+    return this._items!;
```
Where is this used?

iantrich replied:
The return? TypeScript complained if I didn't return a Promise from the async function

balloob replied:
`this._items` is not a promise, so you're not returning a promise. If TypeScript complains, you might have the type declaration of the function wrong

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,236 @@
+      }
+
+      root.value = "";
+      setTimeout(() => root.focus(), 10);
```
Why wait 10 ms?

iantrich replied:
carry over from panel that I didn't question

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,236 @@
+      this._fetchData()
+    );
+
+    setTimeout(() => ev.target.blur(), 10);
```
Why would we do this? What if the user is not done changing the element?

iantrich replied:
This only fires when we hit enter or lose focus, and this is for when we do the former

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,236 @@
+            tabindex='0'
+          ></paper-checkbox>
+          <paper-item-body>
+            <paper-input
```
Look, adding this is not MVP.

MVP stands for minimum viable product. The least amount of logic to get this card merged. So that would just be a card that renders items and allows checkoxes.

Instead this PR tries to do EVERYTHING . Adding items, clearing items, checking items, editing individual items.

iantrich replied:
I get MVP, but really don't like that idea. I know this has been quite rough going through this; so maybe I'll take a step back and make sure this is truly good to go before making any more commits or perhaps even removing this PR until I have polish.

balloob replied:
You might not like the idea of MVP, but then there are multiple people involved in a PR. 

MVP makes everyones life's easier.

- Smallish initial PR to get basic functionality in place. Easy to see why code is there.
- New PR to add feature X, we see exactly each line that is related to feature X.

balloob replied:
Just to be clear: I am not reviewing this PR any further for now. There are over 250 open PRs in our org right now, many of which are older than this one. Many which are focusing on a single feature or MVP.

iantrich commented:
FYI this is the direction I'm heading

![image](https://user-images.githubusercontent.com/1287159/47629183-b30c9380-db07-11e8-95d5-60ae6fab4fee.png)

iantrich commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,236 @@
+
+    if (this.hass) {
+      this.hass.connection
+        .subscribeEvents(() => this._fetchData(), "shopping_list_updated")
```
👍 works perfectly. thanks.

iantrich commented:
This PR has spiraled. Will come back with a clean MVP later.
